From cf6758bbd13fb3571d2e20d91713776040bb8479 Mon Sep 17 00:00:00 2001
From: Lionel Debieve <lionel.debieve@foss.st.com>
Date: Thu, 17 Nov 2022 18:24:42 +0100
Subject: [PATCH 0504/1695] plat-stm32mp2: open the complete debug access on
 STM32MP2

Open all the debug at boot time for development purpose.
Check the BSEC security state to avoid it on secure close device.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 6da58ae2bba8141d9ee57b203dfed091c2f2ac28 plat-stm32mp2: open the complete debug access on STM32MP2]
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
Signed-off-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Change-Id: I2eba591c7fc4ea39a190efe671ae33af5465de93
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/276337
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/354407
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 core/arch/arm/plat-stm32mp2/main.c | 32 +++++++++++++++++++++++++++++-
 1 file changed, 31 insertions(+), 1 deletion(-)

diff --git a/core/arch/arm/plat-stm32mp2/main.c b/core/arch/arm/plat-stm32mp2/main.c
index f953d81fa..a93204da6 100644
--- a/core/arch/arm/plat-stm32mp2/main.c
+++ b/core/arch/arm/plat-stm32mp2/main.c
@@ -12,6 +12,7 @@
 #include <drivers/stm32_risaf.h>
 #include <drivers/stm32_serc.h>
 #include <drivers/stm32_uart.h>
+#include <drivers/stm32mp_dt_bindings.h>
 #include <initcall.h>
 #include <kernel/abort.h>
 #include <kernel/boot.h>
@@ -185,4 +186,33 @@ void plat_bsec_get_static_cfg(struct stm32_bsec_static_cfg *cfg)
 	cfg->upper_start = STM32MP2_UPPER_OTP_START;
 	cfg->max_id = STM32MP2_OTP_MAX_ID;
 }
-#endif
+
+#define BSEC3_DEBUG_ALL		GENMASK_32(11, 1)
+static TEE_Result init_debug(void)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct clk *dbg_clk = stm32mp_rcc_clock_id_to_clk(CK_SYSDBG);
+	uint32_t state = 0;
+
+	res = stm32_bsec_get_state(&state);
+	if (res)
+		return res;
+
+	if (state != BSEC_STATE_SEC_CLOSED) {
+		if (IS_ENABLED(CFG_WARN_INSECURE))
+			IMSG("WARNING: All debug access are allowed");
+
+		res = stm32_bsec_write_debug_conf(BSEC3_DEBUG_ALL);
+		if (res)
+			panic("Debug configuration failed");
+
+		/* Enable DBG as used to access coprocessor debug registers */
+		assert(dbg_clk);
+		if (clk_enable(dbg_clk))
+			panic("Could not enable debug clock");
+	}
+
+	return res;
+}
+early_init_late(init_debug);
+#endif /* CFG_STM32_BSEC3 */
-- 
2.47.2

